### DISCUSSION TITLE
Automated backup to Google Drive

### DISCUSSION THREAD

**Discussion by PalleRaa** [teslamate-org/teslamate#2385] [Ideas]
Would be very good to get an automated way to do backups in TeslaMate. For many it is very important data that is running on very un-reliable hardware (RPi+crap SD)

I know about the official way to do backups and it works fine, but I might forget that from time to time.

**Comment by marcelwirtz** [comment]
I use the way mentioned here with rclone. Works well and create a backup fully automated every night. I don't think that this needs to be implemented directly in TeslaMate.

[https://www.teslaev.co.uk/how-to-perform-an-automatic-teslamate-backup-to-google-drive/](https://www.teslaev.co.uk/how-to-perform-an-automatic-teslamate-backup-to-google-drive/)

**Reply by PalleRaa** [reply]
I tried that method. But somehow rclone is not allowed to access google drive. I get a access denied when I get to 

```
Copy the link to a browser to create a token, then approve in the Google authentication screen (click Advanced, then click Go to TeslaMate (or whatever you called the service)) then click Allow on the popup, then Allow again for the Confirm your choices screen.
Copy the verification code and paste into your SSH session.
```
in the guide :( But that of course that have nothing to do with teslamate.

**Reply by marcelwirtz** [reply]
Idk if i can help here, i use rclone with a private Nextcloud but i will give it a try. 

Are you still logged in into your google account while copying the url to your browser? Maybe try a private tab in your browser and try the link there.

**Reply by alezanott** [reply]
> I use the way mentioned here with rclone. Works well and create a backup fully automated every night. I don't think that this needs to be implemented directly in TeslaMate.
> 
> https://www.teslaev.co.uk/how-to-perform-an-automatic-teslamate-backup-to-google-drive/

Can this be adapted to a backup on a local NAS?

**Reply by cwanja** [reply]
To some extent, it could. You would just be performing the TeslaMate backup script and then copying (or moving) it to your locally mounted NAS. It is probably _easier_ to a local NAS than gCloud.

**Reply by alezanott** [reply]
> To some extent, it could. You would just be performing the TeslaMate backup script and then copying (or moving) it to your locally mounted NAS. It is probably _easier_ to a local NAS than gCloud.

Thanks @cwanja! I have seen that with rclone I can setup an FTP configuration. Since I don't want to run any script on the NAS but everything on the Raspi this could be an option using the FTP. What do you think about it?

**Reply by cwanja** [reply]
Could do FTP, since it is local. Would still consider SFTP so that it is secured (depending on your risk appetite). You also could just mount a file share (secured or not) and then just move the file from the Pi locally to the file share - it would transfer it similarly.

**Reply by marcelwirtz** [reply]
For a local NAS i would try the approach @cwanja mentioned above with the local mount of a share. Then you only need a script like this which exports the database into a file directly onto your NAS share.

```
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
now=`date +"%A"`
cd /home/pi/tmbackup
sudo /usr/local/bin/docker-compose exec -T database pg_dump -U teslamate teslamate > /mnt/NAS_PATH/tmbackup/teslamate.bck_${now}
```

**Reply by sultantariq** [reply]
> [https://www.teslaev.co.uk/how-to-perform-an-automatic-teslamate-backup-to-google-drive/](https://www.teslaev.co.uk/how-to-perform-an-automatic-teslamate-backup-to-google-drive/)

I have used the same guide but I zip it before transferring it to cloud.

**Comment by ktsenvolt** [comment]
The hardest part is get a Token from Google for gdrive with Raspberry Pi. It works and so far so good. Thanks for the tip.

**Reply by samwhu30** [reply]
Hello, I'm really struggling at this part, are you able to help?
